lib/internal/modules/esm/get_format.js

```diff
@@ -82,6 +84,24 @@ function underNodeModules(url) {
   return StringPrototypeIncludes(url.pathname, '/node_modules/');
 }
 
+/**
+ * Determine whether the given source contains CJS or ESM module syntax.
+ * @param {string} source
+ * @param {URL} url
+ */
+function detectModuleFormat(source, url) {
```

Comment:
The `format` property returned from `resolve` is optional. The addition of this helper seems like it’s becoming required; if we don’t know the format, we’re going to determine it, rather than waiting for `load` to figure it out. This means potentially reading the source from disk twice, unless you preserve what you get from the first read, which `detectModuleFormat` here currently doesn’t. Is there another way to fix the bug where we don’t necessarily read the source in `resolve` if we weren’t already doing so?

Reading the source within `resolve` is also problematic because that _should_ be happening in the `load` hook. If you read file sources here, any custom hooks the user has registered for `load` won’t get applied before detection is run on whatever source is read here. So a TypeScript file would get misidentified as CommonJS because it can’t parse as ESM (or really, because it can’t parse at all) even if after the hooks are applied it’s transpiled into runnable ESM.

Reply:
We should add a test where a registered customization hook changes the format of a module. Like say the module in question is

```js
import { version } from 'node:process'
```

And a custom hook is registered that rewrites it to
```
const { version } = require('node:process')
```

We should somehow verify that it got detected as CommonJS. Maybe a second line in the module is something like `console.log(globalThis.require)` and we confirm that it prints something with `Function: require` (CommonJS) and not `undefined` (ESM).

So then the question becomes, how do we achieve this? It feels like we need to put the detection later in the process, after `load` has run?

Reply:
I don't think I have an answer to this scenario. If resolve detects one module type but then load changes the source code from one valid type (module) to another valid type (cjs) I don't think there's a valid way to make any assumptions during resolve.

Reply:
> I don’t think I have an answer to this scenario.

I mean, we kind of have to solve this in order to land this PR, don’t we? Otherwise this is a change to the hooks API, which might be changing anyway; but I don’t think that one of the changes we want to make to it is that `format` is now required to be returned from `resolve`. For hooks that determine format based on source, it’s obviously easier to do so within `load` where you have the source.

So I guess this is the question: can we somehow preserve the ability for `format` to be optional from `resolve`?

Reply:
@GeoffreyBooth: thanks for looking into it and for adding the mentioned test. One thing for my understanding. In this case the `format` as determined by the `resolve` will be changed by the `load`. At the end of the operation this will have the correct value (ie `commonjs` because the hook transformed the source during `load`) but after `resolve` it was `module`. Not sure how often does this happen and if it is per design allowed. If it is, would it make sense to add a resolve hook to the new test and verify that it was auto-detected to `module` and have the rest checking for `commonjs` after load like it already is with the test you just added?